numpy/_core/src/multiarray/stringdtype/casts.c

```diff
@@ -0,0 +1,1941 @@
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+#define _UMATHMODULE
+
+#include "numpy/ndarraytypes.h"
+#include "numpy/arrayobject.h"
+#include "numpy/halffloat.h"
+#include "numpy/npy_math.h"
+#include "numpy/ufuncobject.h"
+
+#include "common.h"
+#include "numpyos.h"
+#include "umathmodule.h"
+#include "gil_utils.h"
+#include "static_string.h"
+#include "dtypemeta.h"
+#include "dtype.h"
+#include "utf8_utils.h"
+
+#define ANY_TO_STRING_RESOLVE_DESCRIPTORS(safety)                              \
+        static NPY_CASTING any_to_string_##safety##_resolve_descriptors(       \
+                PyObject *NPY_UNUSED(self),                                    \
+                PyArray_DTypeMeta *NPY_UNUSED(dtypes[2]),                      \
+                PyArray_Descr *given_descrs[2], PyArray_Descr *loop_descrs[2], \
+                npy_intp *NPY_UNUSED(view_offset))                             \
+        {                                                                      \
+            if (given_descrs[1] == NULL) {                                     \
+                PyArray_Descr *new =                                           \
+                        (PyArray_Descr *)new_stringdtype_instance(NULL, 1);    \
+                if (new == NULL) {                                             \
+                    return (NPY_CASTING)-1;                                    \
+                }                                                              \
+                loop_descrs[1] = new;                                          \
+            }                                                                  \
+            else {                                                             \
+                Py_INCREF(given_descrs[1]);                                    \
+                loop_descrs[1] = given_descrs[1];                              \
+            }                                                                  \
+                                                                               \
+            Py_INCREF(given_descrs[0]);                                        \
+            loop_descrs[0] = given_descrs[0];                                  \
+                                                                               \
+            return NPY_##safety##_CASTING;                                     \
+        }
+
+ANY_TO_STRING_RESOLVE_DESCRIPTORS(SAFE)
```

Comment:
The two implementations differ in the return value - either `NPY_SAFE_CASTING` or `NPY_UNSAFE_CASTING`, indicating that the cast is safe or unsafe, respectively.

I had it set up this way to match the fixed-width strings, but I just looked again at the logic here, and I think I have it backwards compared with fixed-width strings. We seem to always say casting from another type to string is safe, but the other way is unsafe:

```
>>> np.array([1, 2, 3]).astype(str, casting='safe')
array(['1', '2', '3'], dtype='<U21')
>>> np.array([1, 2, 3]).astype(str, casting='safe').astype(np.int64, casting='safe')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Cannot cast array data from dtype('<U21') to dtype('int64') according to the rule 'safe'
```

And I seem to have set things up backwards for StringDType:

```
>>> np.array([1, 2, 3]).astype(np.dtypes.StringDType(), casting='safe')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Cannot cast array data from dtype('int64') to StringDType() according to the rule 'safe'
>>> np.array([1, 2, 3]).astype(np.dtypes.StringDType(), casting='unsafe')
array(['1', '2', '3'], dtype=StringDType())
```

@seberg do you happen to know why the unicode dtype has its casting safety set up like this?

Reply:
I am not sure I particularly like the logic, predates me.  But I think the logic that somewhat matches is "will roundtrip if you do" `x.astype(other).astype(x.dtype)`.

Or in other words, the int to string can't lose information, string to int obviously can.  (I do not really like considering it safe either way, but maybe there should be another cast level, only that having too many is also just tedious).

(Let me know if I am missig the point with that)

Reply:
I audited the casting safety. Now all the casts from numeric dtypes and datetimes are `SAFE` and casts to those types are `UNSAFE`. I also made casts from the fixed-width strings, bytes, and void safe (because there won't ever be truncation) and from `SAME_KIND`.